builder/vsphere/supervisor/step_create_source.go

```diff
@@ -0,0 +1,306 @@
+	StateKeyVMMetadataSecretCreated = "vm_metadata_secret_created"
+)
+
+type CreateSourceConfig struct {
+	// Name of the the source virtual machine (VM) image.
+	ImageName string `mapstructure:"image_name" required:"true"`
+	// Name of the VM class that describes virtual hardware settings.
+	ClassName string `mapstructure:"class_name" required:"true"`
+	// Name of the storage class that configures storage-related attributes.
+	StorageClass string `mapstructure:"storage_class" required:"true"`
+
+	// Name of the source VM. Defaults to `packer-vsphere-supervisor-built-source`.
+	SourceName string `mapstructure:"source_name"`
+	// Name of the network type to attach to the source VM's network interface. Defaults to empty.
+	NetworkType string `mapstructure:"network_type"`
+	// Name of the network to attach to the source VM's network interface. Defaults to empty.
+	NetworkName string `mapstructure:"network_name"`
+	// Preserve the created objects even after importing them to the vSphere endpoint. Defaults to `false`.
+	KeepInputArtifact bool `mapstructure:"keep_input_artifact"`
+}
+
+func (c *CreateSourceConfig) Prepare() []error {
+	var errs []error
+
+	if c.ImageName == "" {
+		errs = append(errs, fmt.Errorf("'image_name' is required for creating the source VM"))
+	}
+	if c.ClassName == "" {
+		errs = append(errs, fmt.Errorf("'class_name' is required for creating the source VM"))
+	}
+	if c.StorageClass == "" {
+		errs = append(errs, fmt.Errorf("'storage_class' is required for creating the source VM"))
+	}
+
+	if c.SourceName == "" {
+		c.SourceName = DefaultSourceName
+	}
+
+	return errs
+}
+
+type StepCreateSource struct {
+	Config             *CreateSourceConfig
+	CommunicatorConfig *communicator.Config
+
+	Namespace  string
+	KubeClient client.Client
+}
+
+func (s *StepCreateSource) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	logger := state.Get("logger").(*PackerLogger)
+	logger.Info("Creating required source objects in Supervisor cluster...")
+
+	var err error
+	defer func() {
+		if err != nil {
+			state.Put("error", err)
+		}
+	}()
+
+	if err = s.initStep(state); err != nil {
+		return multistep.ActionHalt
+	}
+
+	if err = s.createVMMetadataSecret(ctx, logger); err != nil {
+		return multistep.ActionHalt
+	}
+	state.Put(StateKeyVMMetadataSecretCreated, true)
+
+	if err = s.createVM(ctx, logger); err != nil {
+		return multistep.ActionHalt
+	}
+	state.Put(StateKeyVMCreated, true)
+
+	if err = s.createVMService(ctx, logger); err != nil {
+		return multistep.ActionHalt
+	}
+	state.Put(StateKeyVMServiceCreated, true)
+
+	// Make the source name retrievable in later step.
+	state.Put(StateKeySourceName, s.Config.SourceName)
+
+	logger.Info("Finished creating all required source objects in Supervisor cluster")
+	return multistep.ActionContinue
+}
+
+func (s *StepCreateSource) Cleanup(state multistep.StateBag) {
+	logger := state.Get("logger").(*PackerLogger)
+
+	if s.Config.KeepInputArtifact {
```

Comment:
If there is an error in `Run(...)` maybe after the vm has been created for example in the `s.createVMService` do we still want to honor the KeepInputArtifact option and not clean up? 

I think it there is an error in the Run and a user sets this argument  it would be on them to clean any dangling resources. Is that is not what you expect here we may want to check that if there is no error in state and this flag is set then don't clean up. 

CleanUp is called as part of the Packer builder teardown for both successful and failed step Runs to clean up any artifacts that may have been created in the running the step.

Reply:
I consider that it's still valuable to honor the `KeepInputArtifact` even when an error occurs. By keeping the source objects, Packer users would be able to look into those objects themselves to investigate the problems deeper (e.g. checking out K8s events from those objects inside the cluster). Otherwise, the only option I can see to achieve this would be setting `-on-error=ask` in running Packer build command, which I don't think it's as common to users as just setting this config.

Reply:
Exactly. I am going to resolve this conversation since we are going to keep it as is for now.